photon-diagnostics/src/main/scala/com/linkedin/photon/ml/Evaluation.scala

```diff
@@ -129,49 +130,46 @@ object Evaluation extends Logging {
 
   // See https://en.wikipedia.org/wiki/Poisson_regression
   private def poissonRegressionLogLikelihood(labeled: RDD[LabeledPoint], model: PoissonRegressionModel): Double = {
-    val logLikelihoods = labeled.map(sample => {
+
+    val logLikelihoods = labeled.map { sample =>
       // Compute the log likelihoods
       val y = sample.label
       val wTx = sample.computeMargin(model.coefficients.means)
       val numeratorLog = y * wTx - math.exp(wTx)
       val denominatorLog = Gamma.logGamma(1.0 + y) // y! = Gamma(y + 1)
-      numeratorLog - denominatorLog
-    })
 
-    averageRDD(logLikelihoods)
+      (numeratorLog - denominatorLog, 1)
+    }
+
+    averageLogLikelihoodRDD(logLikelihoods)
   }
 
   // See https://en.wikipedia.org/wiki/Logistic_regression
   private def logisticRegressionLogLikelihood(scoreAndLabel: RDD[(Double, Double)]): Double = {
-    val logLikelihood = scoreAndLabel.map{ case (score, label) =>
+
+    val logLikelihoods = scoreAndLabel.map { case (score, label) =>
       val logP = if (score > EPSILON) math.log(score) else math.log(EPSILON)
-      val log1mP = if (score > 1 - EPSILON) math.log1p(1 - EPSILON) else math.log1p(-score)
-      val result =  label * logP + (1.0 - label) * log1mP
+      val log1mP = if (score > 1 - EPSILON) math.log(EPSILON) else math.log1p(-score)
+      val result = label * logP + (1.0 - label) * log1mP
+
       assert(!result.isInfinite && !result.isNaN, s"label = $label, score = $score, result is not finite")
-      result
+
+      (result, 1)
     }
 
-    averageRDD(logLikelihood)
+    averageLogLikelihoodRDD(logLikelihoods)
   }
 
-  private def averageRDD(toAverage: RDD[Double]): Double = {
-    toAverage.mapPartitions(toAverage => {
-      // Compute per-partition partial mean
-      var count = 0
-      var mean = 0.0
-      toAverage.foreach(x => {
-        count += 1
-        mean += (x - mean) / count
-      })
-      Seq((count, mean)).iterator
-    }).reduce((a, b) => {
-      // Aggregate per-partition means
-      val (countA, meanA) = a
-      val (countB, meanB) = b
-      val newCount = countA + countB
-      val newMean = meanA + (meanB * countB) / newCount
-      (newCount, newMean)
-    })._2
+  private def averageLogLikelihoodRDD(logLikelihoods: RDD[(Double, Int)]): Double = {
+
+    val (logLikelihood, count) = logLikelihoods.reduce { (a, b) =>
+      val (logLikelihoodA, countA) = a
+      val (logLikelihoodB, countB) = b
+
+      (logLikelihoodA + logLikelihoodB, countA + countB)
+    }
+
+    logLikelihood - math.log(count)
```

Comment:
Where is this formula coming from?

Reply:
From a misunderstanding I had - changing it